### DISCUSSION TITLE
Importing list of keywords for parser for case-insensitive language

### DISCUSSION THREAD

**Discussion by katawful** [tree-sitter/tree-sitter#1063] [Q&A]
I am working on creating a parser for a game scripting language (TES4 Oblivion), and I'm having trouble on finding a way to add keywords that doesn't clog up the grammer.js file but also easily allows the keywords to be case insensitive. Normally I would do something like:

```javascript
object: $ => choice(
    /[iI][tT][eE][mM]/,
...
),
```
which works fine for something like a conditional or loop, but there are thousands of specific keywords that an end user can use, and I don't want to leave them out of parser outright since I would like to be able to detect basic syntax errors. I am using JS as the binding language, does anyone have any suggestions? I wasn't able to find a similar language that's in this situation (lots of keywords and also case-insensitive)

**Comment by theHamsta** [comment]
See how they solved it for Fortran: https://github.com/stadelmanma/tree-sitter-fortran

**Reply by katawful** [reply]
The case-insensitive function it uses is a lot more elegant than my solution, but it doesn't have a lot of keywords. There's only ever going to be ~3000 functions for this language I'm working on so I'd like to include them if at all possible. I don't want to have 90% of grammar.js dedicated to these keywords, it would make it very unwieldly to work on

**Comment by maxbrunsfeld** [comment]
Are these words really keywords? That is, do they each have unique syntactic forms associated with them?

Just to check - are they actually just built-in functions or variables? If so, Those shouldn’t  be modeled in the grammar, they should be identified at a higher level of your “stack”.

**Reply by katawful** [reply]
They are built in functions, they're not based on a user supplied library. The problem with just treating them as objects like variables is that they have many different ways of being used, and can appear anywhere functionally, so something like this would be difficult to parse otherwise:

```
if variable1 == variable2
if variable1 == function
if variable1 == (function variable2)
if variable1 == (reference.function)
if variable1 == (namespace.variable2)
```
I need a way to distinguish each of these uniquely, and I cannot think of a way to do it without keywords. But the problem is the ~3000 of said functions available to the user

**Reply by maxbrunsfeld** [reply]
Ok, so distinguishing these should not be the parser's job. Just parse them all as `identifier` (or something) in your grammar. Then, when analyzing your code, check the text of the `identifier` tokens, and handle built-in identifiers based on their text.

**Reply by katawful** [reply]
Ok, that makes sense. Thank you!
